test/local/safe-url.js

```diff
@@ -0,0 +1,89 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const assert = require('insist')
+const mocks = require('../mocks')
+
+describe('require:', () => {
+  let log, SafeUrl
+
+  beforeEach(() => {
+    log = mocks.mockLog()
+    SafeUrl = require('../../lib/safe-url')(log)
+  })
+
+  it('returned the expected interface', () => {
+    assert.equal(typeof SafeUrl, 'function')
+    assert.equal(SafeUrl.length, 2)
+  })
+
+  describe('instantiate:', () => {
+    let safeUrl
+
+    beforeEach(() => {
+      safeUrl = new SafeUrl('/foo/:bar', 'baz')
+    })
+
+    it('returned the expected interface', () => {
+      assert.equal(typeof safeUrl.render, 'function')
+      assert.equal(safeUrl.render.length, 0)
+    })
+
+    it('interpolates correctly', () => {
+      assert.equal(safeUrl.render({ bar: 'wibble' }), '/foo/wibble')
+    })
+
+    it('did not call log.error', () => {
+      assert.equal(log.error.callCount, 0)
+    })
+
+    it('logs an error and throws when param is missing', () => {
+      assert.throws(() => safeUrl.render({}))
+      assert.equal(log.error.callCount, 1)
+      assert.deepEqual(log.error.args[0][0], {
+        op: 'safeUrl.mismatch',
+        keys: [],
+        expected: [ 'bar' ],
+        caller: 'baz'
+      })
+    })
+
+    it('logs an error and throws when param has wrong key', () => {
+      assert.throws(() => safeUrl.render({ qux: 'wibble' }))
+      assert.equal(log.error.callCount, 1)
+      assert.deepEqual(log.error.args[0][0], {
+        op: 'safeUrl.unexpected',
+        key: 'qux',
+        expected: [ 'bar' ],
+        caller: 'baz'
+      })
+    })
+
+    it('logs an error and throws when param seems unsafe', () => {
```

Comment:
Should we add some more test cases for errors? Maybe other special characters, undefined, nested params.